### PR TITLE
Bytecode

### PR DESCRIPTION
**Pull request by sjorn3**
Added a rather large module containing everything relating to bytecode.

A limited but working set has been implemented, including functions, memory and exports. Internal function names are lost but otherwise I've gotten the fib/pow/this example to work when called from wasm-ffi and then be loaded in as a module again. Once the Interpreter is merged in I'll be able to write some proper tests for it.

I think a better way to add the allocate/deallocate functions that I currently have stored in `base.wast` would be to have it as a module that is added to as opposed to a text file that is added to, which would allow for some niceties such as interpreting these base functions and being able to output complete modules as bytecode. 

Having base saved as bytecode would make that a lot easier, since the syntax of wast seems to be quite lax making the parser best suited just to our own output (without a perhaps complex rewrite) whilst `bytecode.jl` can read in compiled code from any source (eventually) with ease.

My intentions for what to do next with this are:

  * Improve support for memory operations, perhaps as a separate type of operation.
  * Add support for them to bytecode input/output and the interpreter
  * Create a compiled form of base.wast and (with any luck) interpret it.
  * Also add general improvements such as being able to store the internal function names properly and just generally [increasing coverage](https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md)

After that's done it should be much simpler to add new features such as structs without needing to go over to js to test everything.

### CODE REVIEW

**Comment by sjorn3** [comment]
I didn't realise at first that you could call read with the amount of bytes to read in so I might refactor to make use of that as it should clean up some of the code quite a bit. I'm assuming the implementation is efficient so even if you only read one byte at a time it would still be fast.

File: src/wasm.jl [hunk 1]
```diff
@@ -54,16 +54,24 @@ end
 
 struct Block <: Instruction
   body::Vector{Instruction}
+
+  result::Union{WType, Void}
 end
+Block(body) = Block(body, Void())
```
Comment (MikeInnes) [inline src/wasm.jl]: `Void()` -> nothing

Can you mention why this needs a change to the IR format? Would we need it even without the bytecode writer?

Comment (sjorn3) [reply src/wasm.jl]: To me it seems quite likely we'll need it at some point as you can't have an if block push a value to the stack without it. E.g.:
```
(module
  (func $this (result i64)
    i32.const 0
    if (result i64)
      i64.const 10
    else
      i64.const 12
    end
    return)
)
```
Won't compile through `wast2wasm` without the result mentioned in the if. I think it's reasonable an optimisation might look for a local being set at the end of each branch and restructuring to set the local after the if. Like a ternary operator. Currently the problem hasn't shown itself as blocks have been setting local variables and moving on without adding anything on the stack. (An if with no else can't have a result)

Currently the main reason though is that `base.wast` in the `arrays` branch does have results, and I'd like to be able to read it in with minimal manual reformatting and without building a bytecode ir.

Really whether it stays would depend on the goal of the internal format, which I'd assume is somewhere between coverage and strictly what's required by the compile pipeline.

Comment (MikeInnes) [reply src/wasm.jl]: Yeah, that's fine if it maps directly to WASM features. The other option is to derive the type but storing it is probably a lot simpler. We should definitely be making use of this eventually, in order to make compiled code more compact.

**Comment by MikeInnes** [comment]
LGTM though note the rebase needed due to the interpreter PR.

**Comment by tshort** [comment]
I'm good with it, too. I was wondering how linking was going to be handled. I've looked but haven't found a generic WASM linker. This seems like it'll cover our needs fine.

**Comment by sjorn3** [comment]
Rebased and I've added the fib/pow/this example as a test.

**Comment by sjorn3** [comment]
Most of the different code sections are now supported, though many only minimally to get base.wasm importing. I've also included the work to interpret memory operations and also include them in the data structure as a separate type of operation. It should now be quite simple to compile and interpret simple array functions once the swapping is implemented. The Bytecode-Interpret test shows some array operations functioning.

**Comment by sjorn3** [comment]
An array summing test has been added, in which the function is compiled and used without alteration. Most of the changes are on the Charlotte side. Something very strange starts happening when you try to allocate too many arrays (I tried to make an array of arrays and sum them) and it just doesn't work, so I'm going to see if it's the same story on the javascript side soon. Otherwise something very subtle is going wrong in the interpreter.

**Comment by tshort** [comment]
@sjorn3, exciting stuff! Where are the changes on the Charlotte side? Or, is that still to come?

**Comment by sjorn3** [comment]
yeah it's all on the arrays branch, I'll turn it into a pr now.

**Comment by sjorn3** [comment]
Contained in #19

**Pull request closed without merging**